quickwit-search/src/root.rs

```diff
@@ -48,6 +51,127 @@ fn compute_split_cost(_split_metadata: &SplitMetadata) -> u32 {
     1
 }
 
+#[derive(Clone, Debug)]
+pub struct NodeSearchError {
+    status: tonic::Status,
+    split_ids: Vec<String>,
+}
+
+type SearchResultsByAddr = HashMap<SocketAddr, Result<LeafSearchResult, NodeSearchError>>;
+
+pub async fn execute_search(
+    assigned_leaf_search_jobs: &[(WrappedSearchServiceClient, Vec<Job>)],
+    search_request_with_offset_0: SearchRequest,
+) -> anyhow::Result<SearchResultsByAddr> {
+    // Perform the query phase.
+    let mut result_per_node_addr_futures = HashMap::new();
+
+    // Perform the query phase.
+    for (search_client, jobs) in assigned_leaf_search_jobs.iter() {
+        let leaf_search_request = LeafSearchRequest {
+            search_request: Some(search_request_with_offset_0.clone()),
+            split_ids: jobs.iter().map(|job| job.split.clone()).collect(),
+        };
+
+        debug!(leaf_search_request=?leaf_search_request, grpc_addr=?search_client.grpc_addr(), "Leaf node search.");
+        let mut search_client_clone = search_client.clone();
+        let handle = tokio::spawn(async move {
+            let split_ids = leaf_search_request.split_ids.to_vec();
+            search_client_clone
+                .client()
+                .leaf_search(leaf_search_request)
+                .await
+                .map(|resp| resp.into_inner())
+                .map_err(|tonic| NodeSearchError {
+                    status: tonic,
+                    split_ids,
+                })
+        });
+        result_per_node_addr_futures.insert(search_client.grpc_addr(), handle);
+    }
+    let mut result_per_node_addr = HashMap::new();
+    for (addr, search_result) in result_per_node_addr_futures {
+        ////< An error here means that the tokio task panicked... Not that the grpc erorred.
+        //< The latter is handled later.
+        result_per_node_addr.insert(addr, search_result.await?);
+    }
+
+    Ok(result_per_node_addr)
+}
+
+pub struct ErrorRetries {
+    #[allow(dead_code)]
+    retry_split_ids: Vec<String>,
+    #[allow(dead_code)]
+    nodes_to_use: Vec<SocketAddr>,
+}
+
+/// There are different information which could be useful
+/// Scenario: Multiple requests on a node
+/// - Are all requests of one node failing? -> Don't consider this node as alternative of failing
+/// requests
+/// - Is the node ok, but just the split failing?
+/// - Did all requests fail? (Should we retry in that case?)
+pub fn analyze_errors(search_result: &SearchResultsByAddr) -> Option<ErrorRetries> {
+    let contains_errors = search_result.values().any(|res| {
+        res.is_err()
+            || res
+                .as_ref()
+                .map(|ok_res| ok_res.failed_requests.is_empty())
+                .unwrap_or(false)
+    });
+    if !contains_errors {
+        return None;
+    }
+    let complete_failure_nodes: Vec<_> = search_result
+        .iter()
+        .filter(|(_addr, result)| {
+            result.is_err()
```

Comment:
```
if let Ok(ok_res) = {
    ok_res.failed_requests.len() as u64 == ok_res.aggregated_results)
} else {
   false
}
```

Reply:
Ideally I would actually have an external helper function that returns an 

enum with nice names.

```
enum LeafOutcome {
    Success,
    PartialFailure,
    TotalFailure
}

fn classify_leaf_failures(leaf_result: Result<...>) -> LeafOutcome {
   match leaf_result {
        // ...
   }
}
```